File: Chapter04(Infrared)/IRChat_Desktop_CS/IRChat_Desktop_CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using System.Net;
using System.IO;
using System.Net.Sockets;
using InTheHand.Net.Sockets;

namespace IRChat_Desktop_CS
{
    public partial class Form1 : Form
    {
        //---define the constants---
        const int MAX_MESSAGE_SIZE = 1024;
        const int MAX_TRIES = 3;

        //---define the member variables---
        private string ServiceName = "default";
                        
        public Form1()
        {
            InitializeComponent();
        }

        private void btnSend_Click(object sender, EventArgs e)
        {
            btnSend.Enabled = false;
            SendMessage(MAX_TRIES, txtMessage.Text);
            btnSend.Enabled = true;
            txtMessage.Text = string.Empty;
            txtMessage.Focus();
        }

        private void SendMessage(int NumRetries, string str)
        {
           IrDAClient client = null;
            int CurrentTries = 0;
            //---try to establish a connection---
            do
            {
                try
                {
                    client = new IrDAClient(ServiceName);
                }
                catch (Exception se)
                {
                    if ((CurrentTries >= NumRetries))
                    {
                        throw se;
                    }
                }
                CurrentTries = CurrentTries + 1;
            } while (client == null & CurrentTries < NumRetries);

            //---timeout occurred---
            if ((client == null))
            {
                txtMessagesArchive.BeginInvoke(new myDelegate(UpdateStatus),
                   new object[] { "Error establishing contact" });
                return;
            }

            //---send the message over a stream object---
            System.IO.Stream stream = null;
            try
            {
                stream = client.GetStream();
                stream.Write(System.Text.ASCIIEncoding.ASCII.GetBytes(str), 0, str.Length);

                //---update the status bar---
                txtMessagesArchive.BeginInvoke(new myDelegate(UpdateStatus),
                   new object[] { "Message sent!" });

                //---display the message that was sent---
                txtMessagesArchive.Text = str + Environment.NewLine + txtMessagesArchive.Text;
            }
            catch (Exception e)
            {
                txtMessagesArchive.BeginInvoke(new myDelegate(UpdateStatus),
                   new object[] { "Error sending message." });
            }
            finally
            {
                if ((!(stream == null)))
                {
                    stream.Close();
                }
                if ((!(client == null)))
                {
                    client.Close();
                }
            }
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            txtMessage.Focus();

            //---receive incoming messages as a separate thread---
            System.Threading.Thread t1;
            t1 = new System.Threading.Thread(ReceiveLoop);
            t1.Start();
        }

        public void ReceiveLoop()
        {
            string strReceived;
            strReceived = ReceiveMessage();

            //---keep on listening for new message---
            while (true)
            {
                if (strReceived != string.Empty)
                {
                    txtMessagesArchive.BeginInvoke(
                       new myDelegate(UpdateTextBox),
                       new object[] { strReceived });
                }
                strReceived = ReceiveMessage();
            }
        }

        private string ReceiveMessage()
        {
            int bytesRead = 0;
            IrDAListener listener = new IrDAListener(ServiceName);
            IrDAClient client = null;
            System.IO.Stream stream = null;
            byte[] Buffer = new byte[MAX_MESSAGE_SIZE - 1];
            string str = string.Empty;
            try
            {
                listener.Start();

                //---blocking call---
                client = listener.AcceptIrDAClient();
                stream = client.GetStream();
                bytesRead = stream.Read(Buffer, 0, Buffer.Length);

                //---format the received message---
                str = ">" + System.Text.ASCIIEncoding.ASCII.GetString(Buffer, 0, bytesRead);
            }
            catch (SocketException ex)
            {
                //---ignore error---
            }
            catch (Exception e)
            {
                txtMessagesArchive.BeginInvoke(new myDelegate(UpdateStatus),
                   new object[] { e.ToString() });
            }
            finally
            {
                if ((!(stream == null)))
                {
                    stream.Close();
                }
                if ((!(client == null)))
                {
                    client.Close();
                }
                listener.Stop();
            }
            return str;
        }
        
        private delegate void myDelegate(string str);
        private void UpdateTextBox(string str)
        {
            //---delegate to update the textbox control---
            txtMessagesArchive.Text = str + Environment.NewLine + txtMessagesArchive.Text;
        }
        private void UpdateStatus(string str)
        {
            //---delegate to update the statusbar control---
            ToolStripStatusLabel1.Text = str;
        }
    }
}

File: Chapter04(Infrared)/IRChat_Desktop/IRChat_Desktop/Form1.vb
Imports System.Net
Imports System.IO
Imports System.Net.Sockets
Imports InTheHand.Net.Sockets

Public Class Form1

    '---define the constants---
    Const MAX_MESSAGE_SIZE As Integer = 1280
    Const MAX_TRIES As Integer = 3

    Private ServiceName As String = "default"

    Private Sub btnSend_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSend.Click
        btnSend.Enabled = False
        SendMessage(MAX_TRIES, txtMessage.Text)
        btnSend.Enabled = True
        txtMessage.Text = String.Empty
        txtMessage.Focus()
    End Sub

    Private Sub SendMessage(ByVal NumRetries As Integer, ByVal str As String)
        Dim client As IrDAClient = Nothing
        Dim CurrentTries As Integer = 0

        '---try to establish a connection---
        Do
            Try
                client = New IrDAClient(ServiceName)
            Catch se As Exception
                If (CurrentTries >= NumRetries) Then
                    Throw se
                End If
            End Try
            CurrentTries = CurrentTries + 1
        Loop While client Is Nothing And CurrentTries < NumRetries

        '---timeout occurred---
        If (client Is Nothing) Then
            'lblStatus.BeginInvoke( _
            txtMessagesArchive.BeginInvoke( _
               New myDelegate(AddressOf UpdateStatus), New Object() _
               {"Error establishing contact"})
            Return
        End If

        '---send the message over a stream object---
        Dim stream As System.IO.Stream = Nothing
        Try
            stream = client.GetStream()
            stream.Write(System.Text.ASCIIEncoding.ASCII.GetBytes(str), 0, str.Length)
            '---update the status bar---
            txtMessagesArchive.BeginInvoke( _
               New myDelegate(AddressOf UpdateStatus), New Object() _
               {"Message sent!"})
            '---display the message that was sent---
            txtMessagesArchive.Text = str & vbCrLf & txtMessagesArchive.Text
        Catch e As Exception
            txtMessagesArchive.BeginInvoke( _
               New myDelegate(AddressOf UpdateStatus), New Object() _
               {"Error sending message."})
        Finally
            If (Not stream Is Nothing) Then stream.Close()
            If (Not client Is Nothing) Then client.Close()
        End Try
    End Sub

    Private Function ReceiveMessage() As String
        Dim bytesRead As Integer = 0
        Dim listener As IrDAListener = New IrDAListener(ServiceName)
        Dim client As IrDAClient = Nothing
        Dim stream As System.IO.Stream = Nothing
        Dim Buffer(MAX_MESSAGE_SIZE - 1) As Byte
        Dim str As String = String.Empty
        Try
            listener.Start()
            client = listener.AcceptIrDAClient()  '---blocking call---
            stream = client.GetStream()
            bytesRead = stream.Read(Buffer, 0, Buffer.Length)
            '---display the received message---
            str = ">" & System.Text.ASCIIEncoding.ASCII.GetString(Buffer, 0, bytesRead)
        Catch ex As SocketException
            '---ignore error---
        Catch e As Exception
            txtMessagesArchive.BeginInvoke( _
               New myDelegate(AddressOf UpdateStatus), New Object() _
               {e.ToString})
        Finally
            If (Not stream Is Nothing) Then stream.Close()
            If (Not client Is Nothing) Then client.Close()
            listener.Stop()
        End Try
        Return str
    End Function

    Public Sub ReceiveLoop()
        Dim strReceived As String
        strReceived = ReceiveMessage()
        '---keep on listening for new message
        While True
            If strReceived <> String.Empty Then
                txtMessagesArchive.BeginInvoke( _
                New myDelegate(AddressOf UpdateTextBox), New Object() {strReceived})
            End If
            strReceived = ReceiveMessage()
        End While
    End Sub

    Private Delegate Sub myDelegate(ByVal str As String)
    Private Sub UpdateTextBox(ByVal str As String)
        '---delegate to update the textbox control
        txtMessagesArchive.Text = str & vbCrLf & txtMessagesArchive.Text
    End Sub

    Private Sub UpdateStatus(ByVal str As String)
        '---delegate to update the statusbar control
        ToolStripStatusLabel1.Text = str
    End Sub

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        txtMessage.Focus()
        '---receive incoming messages as a separate thread---
        Dim t1 As System.Threading.Thread
        t1 = New Threading.Thread(AddressOf ReceiveLoop)
        t1.Start()
    End Sub
End Class
